### DISCUSSION TITLE
How to save and load  best DeepAR model ?

### DISCUSSION THREAD

**Discussion by karndeepsingh** [awslabs/gluonts#1762] [Q&A]
Hi,
I have been training DeepAR model for timeseries forecasting. I want to know how we can save best weights or best trained model and how it can be loaded back and used for prediction?

Please, provide some help on this.
Thanks

**Comment by dai-ichiro** [comment]
```
import pickle

with open('gluonts_model', 'wb') as f:
    pickle.dump(predictor, f)

with open('gluonts_model', 'rb') as f:
    predictor = pickle.load(f)
```

**Reply by karndeepsingh** [reply]
How could I save best weights? Does predictor object  saves best checkpoints or the latest checkpoints?

**Comment by jaheba** [comment]
Hi,

you should be able to call `model.serialize(<path>)`:

https://github.com/awslabs/gluon-ts/blob/569fcc779c66a81c843857cca76aa1f07a8ea0fc/src/gluonts/model/predictor.py#L85

To load the model, you just call `Predictor.deserialize(<path>)`.

**Reply by vmajor** [reply]
`Predictor.deserialize(<path>)` generates this error: `FileNotFoundError: [Errno 2] No such file or directory: 'lightning_logs/version_83/checkpoints/type.txt'`

`type.txt` is also never saved during training so I do not even know what it is and where it might be found. Thus please excuse my confusion but please let me know how can I load the saved checkpoint generated using a torch based model.

**Reply by jaheba** [reply]
This looks like checkpoints generated by lightning, not by GluonTS through `serialize`. Can you share a bit more how you train a model and how you intend to load it?

**Reply by vmajor** [reply]
I am using the pytorch version of TFT and am doing this to train it. Checkpoints are saved automatically as a function of Lightning:

```
from gluonts.torch.model.tft import TemporalFusionTransformerEstimator

  # Train a TFT model on the data
   # Toggle comment when enabling Optuna
    estimator = TemporalFusionTransformerEstimator(
        prediction_length=prediction_length,
        freq=freq,
        trainer_kwargs={"max_epochs": 25},
        context_length=8,
        dropout_rate=0.40410747064575153,
        hidden_dim=16,
        num_heads=8,
    )

    #model = estimator.train(train_dataset, val_dataset, ckpt_path='trained_models/tft_model.ckpt')
    model = estimator.train(train_dataset, val_dataset, num_workers=11)
```

**Reply by SummitStudiosDev** [reply]
did you ever figure out how to load from a pytorch lightning checkpoint by chance?

**Reply by vmajor** [reply]
I do not remember, sorry. I did manage to find a way to load from checkpoint but I vaguely remember I used the built in method from gluonts which is in fact the pytorch method. I am sorry I am unable to help more. I abandoned this approach because I found that I do not actually want to load pretrained models in my use case so this solution and related code was not retained.

**Reply by SummitStudiosDev** [reply]
that's ok! i just ended up retraining it and using the intended method of serializing it

**Reply by Somendratripathi** [reply]
Using train_model method returns the best model which you could serealize. 

https://ts.gluon.ai/stable/_modules/gluonts/torch/model/deep_npts/_estimator.html#DeepNPTSEstimator.train_model
